File: CS/ReportServerWinForms/MainForm.cs
using System;
using System.IO;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.DocumentServices.ServiceModel;
using DevExpress.DocumentServices.ServiceModel.DataContracts;
using DevExpress.ReportServer.ServiceModel.Client;
using DevExpress.ReportServer.ServiceModel.ConnectionProviders;
using DevExpress.ReportServer.ServiceModel.DataContracts;
using DevExpress.XtraPrinting;
using DevExpress.XtraReports.UI;

namespace ReportServerWinFormsClientDemo {
    public partial class MainForm : Form {
        const string ServerAddress = "https://reportserver.devexpress.com";
        readonly ConnectionProvider serverConnection = new GuestConnectionProvider(ServerAddress);

        readonly int ReportId = 1113; //Customer Order History report
        readonly ReportParameter[] reportParameters = new ReportParameter[] {
            new ReportParameter {
                Path = "@CustomerID", 
                Value = "BERGS",
                MultiValue = false
            }};

        public MainForm() {
            InitializeComponent();
        }

        private void export2PDFReportButton_Click(object sender, EventArgs e) {
            if(saveFileDialog1.ShowDialog() != DialogResult.OK) {
                return;
            }
            ExportTo(serverConnection, ReportId, new PdfExportOptions(), reportParameters, x => File.WriteAllBytes(saveFileDialog1.FileName, x));
        }

        private void printReportButton_Click(object sender, EventArgs e) {
            ExportTo(serverConnection, ReportId, new NativeFormatOptions(), reportParameters, x => PrintReportFromPrnx(x));
        }

        void PrintReportFromPrnx(byte[] prnx) {
            using(XtraReport report = new XtraReport()) {
                using(MemoryStream stream = new MemoryStream(prnx)) {
                    report.PrintingSystem.LoadDocument(stream);
                }
                report.Print("Microsoft XPS Document Writer"); //specify a printer name here
            }
        }

        void ExportTo(ConnectionProvider serverConnection, int reportId, ExportOptionsBase exportOptions, ReportParameter[] parameters, Action<byte[]> action) {
            printReportButton.Enabled = false;
            export2PDFReportButton.Enabled = false;
            splashScreenManager1.ShowWaitForm();
            serverConnection
                .ConnectAsync()
                .ContinueWith(t => {
                    IReportServerClient client = t.Result;
                    return Task.Factory.ExportReportAsync(client, new ReportIdentity(reportId), exportOptions, parameters, null);
                }).Unwrap()
                .ContinueWith(t => {
                    splashScreenManager1.CloseWaitForm();
                    printReportButton.Enabled = true;
                    export2PDFReportButton.Enabled = true;
                    try {
                        if(t.IsFaulted) {
                            throw new Exception(t.Exception.Flatten().InnerException.Message);
                        }
                        action(t.Result);
                    } catch(Exception e) {
                        MessageBox.Show(e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    }
                }, TaskScheduler.FromCurrentSynchronizationContext());
        }
    }
}

File: VB/ReportServerWinForms/MainForm.vb
Imports System
Imports System.IO
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.DocumentServices.ServiceModel
Imports DevExpress.DocumentServices.ServiceModel.DataContracts
Imports DevExpress.ReportServer.ServiceModel.Client
Imports DevExpress.ReportServer.ServiceModel.ConnectionProviders
Imports DevExpress.ReportServer.ServiceModel.DataContracts
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraReports.UI

Namespace ReportServerWinFormsClientDemo
    Partial Public Class MainForm
        Inherits Form

        Private Const ServerAddress As String = "https://reportserver.devexpress.com"
        Private ReadOnly serverConnection As ConnectionProvider = New GuestConnectionProvider(ServerAddress)

        Private ReadOnly ReportId As Integer = 1113 'Customer Order History report
        Private ReadOnly reportParameters() As ReportParameter = { _
            New ReportParameter With {.Path = "Northwind.CustOrderHist.@CustomerID", .Value = "BERGS", .MultiValue = False} _
        }

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub export2PDFReportButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles export2PDFReportButton.Click
            If saveFileDialog1.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
                Return
            End If
            ExportTo(serverConnection, ReportId, New PdfExportOptions(), reportParameters, Sub(x) File.WriteAllBytes(saveFileDialog1.FileName, x))
        End Sub

        Private Sub printReportButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles printReportButton.Click
            ExportTo(serverConnection, ReportId, New NativeFormatOptions(), reportParameters, Sub(x) PrintReportFromPrnx(x))
        End Sub

        Private Sub PrintReportFromPrnx(ByVal prnx() As Byte)
            Using report As New XtraReport()
                Using stream As New MemoryStream(prnx)
                    report.PrintingSystem.LoadDocument(stream)
                End Using
                report.Print("Microsoft XPS Document Writer") 'specify a printer name here
            End Using
        End Sub

        Private Sub ExportTo(ByVal serverConnection As ConnectionProvider, ByVal reportId As Integer, ByVal exportOptions As ExportOptionsBase, ByVal parameters() As ReportParameter, ByVal action As Action(Of Byte()))
            printReportButton.Enabled = False
            export2PDFReportButton.Enabled = False
            splashScreenManager1.ShowWaitForm()
            serverConnection.ConnectAsync().ContinueWith(Function(t)
                Dim client As IReportServerClient = t.Result
                Return Task.Factory.ExportReportAsync(client, New ReportIdentity(reportId), exportOptions, parameters, Nothing)
            End Function).Unwrap().ContinueWith(Sub(t)
                splashScreenManager1.CloseWaitForm()
                printReportButton.Enabled = True
                export2PDFReportButton.Enabled = True
                Try
                    If t.IsFaulted Then
                        Throw New Exception(t.Exception.Flatten().InnerException.Message)
                    End If
                    action(t.Result)
                Catch e As Exception
                    MessageBox.Show(e.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                End Try
End Sub, TaskScheduler.FromCurrentSynchronizationContext())
        End Sub
    End Class
End Namespace
